pkg/policy/mapstate.go

```diff
@@ -74,24 +79,100 @@ type MapState interface {
 
 // mapState is a state of a policy map.
 type mapState struct {
-	allows mapStateMap
-	denies mapStateMap
+	allows *mapStateMap
+	denies *mapStateMap
+}
+
+type Identities interface {
+	GetNetsLocked(identity.NumericIdentity) []*net.IPNet
 }
 
 // mapStateMap is a convience type representing the actual structure mapping
 // policymap keys to policymap entries.
-type mapStateMap map[Key]MapStateEntry
+type mapStateMap struct {
+	t   bitlpm.Trie[bitlpm.Key[Key], map[uint32]MapStateEntry]
```

Comment:
It is not clear why each trie entry is a map keyed with an id, while the id field in the key is zeroed out on Upsert. At the minimum it would look like the id should also be zeroed out on Lookup. The only rationale I can think of is to exercise the trie lookup by only inserting zeroed identities, but then relying on that ancestor to be found on the non-zeroed lookup? I would appreciate some commentary on the specific use of `bitlpm.Trie` here.

Reply:
I wonder if this is an artifact of `Key` having inconvenient ordering of the fields, where the port we're trying to do a prefix match upon is field 2/4. However, when I compare `Key` with the underlying `policy_key` structure, I don't understand why userspace is defining the key in such an inconvenient format; the datapath doesn't have this prefix matching problem because the fields are ordered in a way that's sane for prefix matching.

Could we simplify this by submitting a separate commit to rearrange `Key`'s fields, then just directly use the full `Key` in the `Trie` in this patch?

Reply:
> At the minimum it would look like the id should also be zeroed out on Lookup

It is, but what do you mean by "look like"?

> I wonder if this is an artifact of Key having inconvenient ordering of the fields

The ordering is not the problem and is something that can be "fixed" easily enough if we decide to change it (the order can be arbitrary and is determined by the methods `CommonPrefix` and `BitValueAt`). Making identity part of the key greatly diminishes the efficiency of the data-structure because it turns every key into its own branch and creates an immense number of intermediate nodes in the structure. A single policy could end up getting split up into hundreds of branches, versus having a policy map to one key with a map of identities. I understand that the datapath can't do this (though maybe we should investigate ways to refactor the datapath code), but the way I've laid it out in userspace will improve lookup/insertion/deletion times.

Reply:
Oh I see, yes for some reason I figured that the bitlpm would order the tree based on the inserted prefix lengths with an arbitrary number of branches at each level rather than being a full binary tree. 

Makes sense and I agree that exploring more efficient implementations in datapath would be interesting :+1: